### PR TITLE
Add support for Neuralforecast

### LINKED ISSUE #1112: NeuralForecast support for forecasting
### Search before asking

- [X] I have searched the EvaDB [issues](https://github.com/georgia-tech-db/eva/issues) and found no similar feature requests.

### Description

`Neuralforecast` brings about support for DL-based methods, that would also help bring about support for multivariate forecasting. It also uses `Ray` for faster hyperparameter optimization.

### Use case

Robust and more accurate predictions, along with multivariate forecasting.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by americast**
Adding support for `neuralforecast`. Fixes #1112.

```sql
DROP TABLE IF EXISTS AirData;

CREATE TABLE AirData (
    unique_id TEXT(30),
    ds TEXT(30),
    y INTEGER);

LOAD CSV 'data/forecasting/air-passengers.csv' INTO AirData;

DROP FUNCTION IF EXISTS Forecast;

CREATE FUNCTION Forecast FROM
(SELECT unique_id, ds, y FROM AirData)
TYPE Forecasting
PREDICT 'y'
HORIZON 12
LIBRARY 'neuralforecast';

SELECT Forecast(12);
```
One quick issue here is that `neuralforecast` needs `horizon` as a parameter while training, unlike `statsforecast`. Thus, a better way to call the UDF would be simply `SELECT Forecast();`, which is currently unsupported. @xzdandy Please let me know your thoughts.

List of stuff yet to be done:

- [x] Incorporate `neuralforecast`
- [x] Fix `HORIZON` redundancy (UPDATE: Being fixed in #1121)
- [x] Reuse model with lower horizon no
- [x] Add support for ~multivariate forecasting~ exogenous variables
- [x] Add tests
- [x] Add docs

### CODE REVIEW

**Comment by xzdandy** [comment]
What is the error message when you try `SELECT Forecast()`?

**Comment by americast** [comment]
> What is the error message when you try `SELECT Forecast()`?

```
    raise UnexpectedToken(token, e.allowed, state=parser_state, token_history=[last_token], terminals_by_name=self.root_lexer.terminals_by_name)
lark.exceptions.UnexpectedToken: Unexpected token Token('RR_BRACKET', ')') at line 1, column 17.
Expected one of: 
        * MAX
        * AVG
        * DECIMAL_LITERAL
        * SUM
        * ONE_DECIMAL
        * MINUS
        * LR_SQ_BRACKET
        * LR_BRACKET
        * COUNT
        * BIT_NOT_OP
        * TRUE
        * FALSE
        * LOCAL_ID
        * STRING_LITERAL
        * TWO_DECIMAL
        * SEGMENT
        * STAR
        * REVERSE_QUOTE_ID
        * PLUS
        * ID
        * MIN
        * FIRST
        * REAL_LITERAL
        * ANYDIM
        * NOT
        * LAST
        * EXCLAMATION_SYMBOL
        * ZERO_DECIMAL
Previous tokens: [Token('LR_BRACKET', '(')]
```

**Comment by jarulraj** [comment]
You can copy paste the grammar file and the query into this IDE: https://www.lark-parser.org/ide/

**Comment by jarulraj** [comment]
Nice job, @americast :)

**Comment by xzdandy** [comment]
Will review after #1209 is merged.

File: evadb/binder/statement_binder.py [hunk 1]
```diff
@@ -126,10 +126,10 @@ def _bind_create_function_statement(self, node: CreateFunctionStatement):
...
                         required_columns.remove(column.name)
-                    else:
-                        raise BinderError(
-                            f"Unexpected column {column.name} found for forecasting function."
-                        )
+                    # else:
+                    #     raise BinderError(
+                    #         f"Unexpected column {column.name} found for forecasting function."
+                    #     )
                 assert (
```
Comment (xzdandy) [inline evadb/binder/statement_binder.py:133]: We can just remove the code

File: evadb/executor/create_function_executor.py [hunk 2]
```diff
@@ -255,32 +269,142 @@ def handle_forecasting_function(self):
             frequency.split("-")[0] if "-" in frequency else frequency
         )  # shortens longer frequencies like Q-DEC
         season_length = season_dict[new_freq] if new_freq in season_dict else 1
-        model = StatsForecast(
-            [model_dict[model_name](season_length=season_length)], freq=new_freq
-        )
+
+        try_to_import_forecast()
```
Comment (xzdandy) [inline evadb/executor/create_function_executor.py:273]: We should not require user to install both nerualforecast and statsforecast. Make them separate is better. 
`try_to_import_statsforecast` and `try_to_import_neuralforecast`.

File: test/unit_tests/binder/test_statement_binder.py [hunk 3]
```diff
@@ -568,50 +568,50 @@ def test_bind_create_function_should_bind_forecast_with_renaming_columns(self):
...
+    #             ("predict", "ma"),
+    #             ("time", "saledate"),
+    #         ]
+    #         binder = StatementBinder(StatementBinderContext(MagicMock()))
+
+    #         with self.assertRaises(BinderError) as cm:
+    #             binder._bind_create_function_statement(create_function_statement)
+
+    #         err_msg = "Unexpected column type found for forecasting function."
+    #         self.assertEqual(str(cm.exception), err_msg)
```
Comment (xzdandy) [inline test/unit_tests/binder/test_statement_binder.py:614]: Please remove

**Comment by xzdandy** [comment]
Great work. Looks perfect. I will also update the notebook and push to this PR. Leave some minors comments to fix. Thanks.

File: docs/source/reference/ai/model-forecasting.rst [hunk 4]
```diff
@@ -47,16 +47,22 @@ EvaDB's default forecast framework is `statsforecast <https://nixtla.github.io/s
...
+   * - AUTO (default: 'F')
+     - The names of columns to be treated as exogenous variables, separated by comma. These columns would be considered for forecasting by the backend only for LIBRARY `neuralforecast`.
+   * - Frequency (default: 'auto')
+     - A string indicating the frequency of the data. The common used ones are D, W, M, Y, which repestively represents day-, week-, month- and year- end frequency. The default value is M. Check `pandas available frequencies <https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases>`_ for all available frequencies. If it is not provided, the frequency is attempted to be determined automatically.
+
+Note: If columns other than the ones required as mentioned above are passed while creating the function, they will be treated as exogenous variables if LIBRARY is `neuralforecast` and the AUTO is set to F. In other situations, they would be ignored.
```
Comment (xzdandy) [inline docs/source/reference/ai/model-forecasting.rst:65]: Is `HORIZON` also required now? Could you also update the query in the documentation, and give an example query for neuralforecast? Thanks!

File: docs/source/reference/ai/model-forecasting.rst [hunk 5]
```diff
@@ -47,16 +47,22 @@ EvaDB's default forecast framework is `statsforecast <https://nixtla.github.io/s
...
+   * - ID (default: 'unique_id')
+     - The name of column that represents an identifier for the series. If relevant column is not found, the whole table is considered as one series of data.
+   * - LIBRARY (default: 'statsforecast')
+     - We can select one of `statsforecast` (default) or `neuralforecast`. `statsforecast` provides access to statistical forecasting methods, while `neuralforecast` gives access to deep-learning based forecasting methods.
+   * - MODEL (default: 'AutoARIMA')
+     - If LIBRARY is `statsforecast`, we can select one of AutoARIMA, AutoCES, AutoETS, AutoTheta. The default is AutoARIMA. Check `Automatic Forecasting <https://nixtla.github.io/statsforecast/src/core/models_intro.html#automatic-forecasting>`_ to learn details about these models. If LIBRARY is `neuralforecast`, we can select one of NHITS or NBEATS. The default is NBEATS. Check `NBEATS docs <https://nixtla.github.io/neuralforecast/models.nbeats.html>`_ for details.
```
Comment (xzdandy) [inline docs/source/reference/ai/model-forecasting.rst:59]: ``statsforecast`` formats better than `statsforecast`. Same for the rest.

File: evadb/functions/forecast.py [hunk 6]
```diff
@@ -44,19 +46,20 @@ def setup(
...
+            forecast_df = self.model.predict()
         forecast_df.reset_index(inplace=True)
         forecast_df = forecast_df.rename(
             columns={
                 "unique_id": self.id_column_rename,
                 "ds": self.time_column_rename,
                 self.model_name: self.predict_column_rename,
             }
-        )
+        )[: self.horizon]
```
Comment (xzdandy) [inline evadb/functions/forecast.py:64]: I found out this not work for statsforecast, when updating the notebook. Statsforecast output `horizon` rows for very unique id.

File: test/integration_tests/long/test_model_forecasting.py [hunk 7]
```diff
@@ -92,6 +127,7 @@ def test_forecast_with_column_rename(self):
                 WHERE bedrooms = 2
             )
             TYPE Forecasting
+            HORIZON 24
```
Comment (xzdandy) [inline test/integration_tests/long/test_model_forecasting.py:130]: This should be 12. The output is 24 because there are two values in the `type` column.

**Comment by xzdandy** [comment]
Updated the forecast notebook. Found out some new issues. Please check. We'd also like to try the neuralforecast in the notebook to see if it addresses several limitations we found with statsforecast. Create #1236 for the follow up work.

File: evadb/functions/forecast.py [hunk 8]
```diff
@@ -44,19 +46,20 @@ def setup(
...
+            forecast_df = self.model.predict()
         forecast_df.reset_index(inplace=True)
         forecast_df = forecast_df.rename(
             columns={
                 "unique_id": self.id_column_rename,
                 "ds": self.time_column_rename,
                 self.model_name: self.predict_column_rename,
             }
-        )
+        )[: self.horizon * forecast_df["unique_id"].nunique()]
```
Comment (xzdandy) [inline evadb/functions/forecast.py:64]: How does `neuralforecast` handle `unique_id` column? Is that similar to `statsforecast`?

**Review by xzdandy — APPROVED** [review]

**Comment by xzdandy** [comment]
The PR looks good. Thanks for the good work! I will update the notebook.

**Pull request merged**